### ISSUE TITLE
[Left to spend] on main page shows incorrect value

### ISSUE THREAD

**Issue by Teonyr** [firefly-iii/firefly-iii#7853]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [X] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [ ] ... there are no open or closed issues that are related to my problem
- [X] ... it's [definitely a Firefly III issue, not me](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

The amount of money "left to spend" shown on the main page is incorrect.

According to the budget page, I should have 1173,20€ left to spend.
But on the main page it shows 6583,20€.

Thanks in advance. :)

### Debug information

Debug information generated at 2023-08-15 11:46:31 Europe/Berlin for Firefly III version **6.0.20**.

| Scope | Version |
| --- | --- |
| Firefly III | 6.0.20 |
| Firefly III API | 2.0.6 |
| Build | 581, 12-08-2023 18:02:44 UTC |
| Base Build | 35, 30-07-2023 03:52:37 UTC |
| DB version | 20 (exp. 20) |
| PHP | `8.2.8` |
| Host | `Linux` |

| System info | Value |
| --- | --- |
| System TZ | Europe/Berlin |
| Browser TZ | Europe/Berlin |
| App environment | local |
| App debug mode | false |
| App cache driver | file |
| App logging | notice, stack |
| Display errors | Off |
| Error reporting | ALL errors |
| Max upload | 67108864 (64 MB) |
| Interface | apache2handler |
| Default language | de_DE |
| Default locale | equal |
| BCscale | 12 |
| DB drivers | mysql, pgsql, sqlite |
| Current driver | pgsql |
| Trusted proxies (.env) | `*` |

| User info | Value |
| --- | --- |
| Login provider | eloquent  |
| User guard | web  |
| Headers | N/A, N/A  |
| Stateful domains |  |
| Session start | 2023-07-01 00:00:00 |
| Session end | 2023-09-30 00:00:00 |
| Session first | 2020-08-01 00:00:00 |
| User ID | 1 |
| User language | de_DE |
| User locale | de_DE |
| Attempt at "de_DE.utf8" | 'de_DE.utf8' |
| Attempt at "de_DE.UTF-8" | 'de_DE.UTF-8' |
| User agent | Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/114.0 |

### Expected behaviour

Should show the correct value of 1173,20€.

### Additional info

After doing some calculations, I found out, that the amount "left to spend" on the main page is the sum of the "budgeted" amount (in my case 5410,00€) and the amount I have left to spend (in my case 1173,20€) which results in: 6583,20€ (which is the amount displayed on the main page).

I already looked at this issue: https://github.com/firefly-iii/firefly-iii/issues/7752 and tried your suggestions there (changing a budget and using /flush page), but they didn't help.

![Budget page](https://github.com/firefly-iii/firefly-iii/assets/75175693/237c6427-86f7-417f-9d5e-7368bb3afa6a)

![Left to spend from main page](https://github.com/firefly-iii/firefly-iii/assets/75175693/4db51ddf-2907-40d1-82d7-424d96107275)

**Comment by JC5** [comment]
I see what you mean, but I can't get it replicated. The demo site and my local installation display the amount just fine.

Since you're on Docker, what happens when you run the `develop` tag?

**Comment by Teonyr** [comment]
Unfortunately, it displays the same incorrect amount on the main page.
I also quickly tried /flush and altering a budget on develop again but doesn't help either.

**Comment by Teonyr** [comment]
Quick follow up:
During the start of the container, I saw this error message in log:

```

  [✓] All piggy bank amounts are positive.
  [✓] All recurring transaction amounts are positive.
[2023-08-15 14:43:03] local.ERROR: Exception is: {"class":"ValueError","errorMessage":"bccomp(): Argument #1 ($num1) is not well-formed","time":"Tue, 15 Aug 2023 14:43:03 +0200","file":"\/var\/www\/html\/app\/Console\/Commands\/Correction\/CorrectAmounts.php","line":250,"code":0,"version":"6.0.20","url":"http:\/\/localhost","userAgent":"Symfony","json":true,"method":"GET"}  
[2023-08-15 14:43:03] local.ERROR: bccomp(): Argument #1 ($num1) is not well-formed {"exception":"[object] (ValueError(code: 0): bccomp(): Argument #1 ($num1) is not well-formed at /var/www/html/app/Console/Commands/Correction/CorrectAmounts.php:250)

[stacktrace]
#0 /var/www/html/app/Console/Commands/Correction/CorrectAmounts.php(250): bccomp()
#1 /var/www/html/app/Console/Commands/Correction/CorrectAmounts.php(70): FireflyIII\\Console\\Commands\\Correction\\CorrectAmounts->fixRuleTriggers()
#2 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(36): FireflyIII\\Console\\Commands\\Correction\\CorrectAmounts->handle()
#3 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Util.php(41): Illuminate\\Container\\BoundMethod::Illuminate\\Container\\{closure}()
#4 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(93): Illuminate\\Container\\Util::unwrapIfClosure()
#5 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(35): Illuminate\\Container\\BoundMethod::callBoundMethod()
#6 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Container.php(662): Illuminate\\Container\\BoundMethod::call()
#7 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php(211): Illuminate\\Container\\Container->call()
#8 /var/www/html/vendor/symfony/console/Command/Command.php(326): Illuminate\\Console\\Command->execute()
#9 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php(180): Symfony\\Component\\Console\\Command\\Command->run()
#10 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Concerns/CallsCommands.php(67): Illuminate\\Console\\Command->run()
#11 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Concerns/CallsCommands.php(28): Illuminate\\Console\\Command->runCommand()
#12 /var/www/html/app/Console/Commands/Correction/CorrectDatabase.php(83): Illuminate\\Console\\Command->call()
#13 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(36): FireflyIII\\Console\\Commands\\Correction\\CorrectDatabase->handle()
#14 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Util.php(41): Illuminate\\Container\\BoundMethod::Illuminate\\Container\\{closure}()
#15 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(93): Illuminate\\Container\\Util::unwrapIfClosure()
#16 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(35): Illuminate\\Container\\BoundMethod::callBoundMethod()
#17 /var/www/html/vendor/laravel/framework/src/Illuminate/Container/Container.php(662): Illuminate\\Container\\BoundMethod::call()
#18 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php(211): Illuminate\\Container\\Container->call()
#19 /var/www/html/vendor/symfony/console/Command/Command.php(326): Illuminate\\Console\\Command->execute()
#20 /var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php(180): Symfony\\Component\\Console\\Command\\Command->run()
#21 /var/www/html/vendor/symfony/console/Application.php(1081): Illuminate\\Console\\Command->run()
#22 /var/www/html/vendor/symfony/console/Application.php(320): Symfony\\Component\\Console\\Application->doRunCommand()
#23 /var/www/html/vendor/symfony/console/Application.php(174): Symfony\\Component\\Console\\Application->doRun()
#24 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(201): Symfony\\Component\\Console\\Application->run()
#25 /var/www/html/artisan(35): Illuminate\\Foundation\\Console\\Kernel->handle()
#26 {main}
"} 
   ValueError 
  bccomp(): Argument #1 ($num1) is not well-formed
  at app/Console/Commands/Correction/CorrectAmounts.php:250
    246▕         $fixed = 0;
    247▕         /** @var RuleTrigger $item */
    248▕         foreach ($set as $item) {
    249▕             // basic check:
  ➜ 250▕             if (-1 === bccomp((string)$item->trigger_value, '0')) {
    251▕                 $fixed++;
    252▕                 $item->trigger_value = app('steam')->positive((string)$item->trigger_value);
    253▕                 $item->save();
    254▕             }
  1   app/Console/Commands/Correction/CorrectAmounts.php:250
  2   app/Console/Commands/Correction/CorrectAmounts.php:70
      FireflyIII\Console\Commands\Correction\CorrectAmounts::fixRuleTriggers()
      Now executing firefly-iii:create-group-memberships
  [✓] Validated group memberships
      Now executing firefly-iii:report-empty-objects

```

**Comment by JC5** [comment]
Alright, I can catch that particular error, but it seems unrelated. In the mean time, I've added debug info to the console. Make sure you have `APP_DEBUG=true` and `APP_LOG_LEVEL=debug`, and run the latest `develop`-tag if you would. It should report it's build 590.

I hope the logs will make it more clear.

By the way, the console may complain about a rule being disabled from now on, please check that out at a later time. It seems one of your rules had an invalid rule trigger.

**Comment by Teonyr** [comment]
The rule in question was my 4,42€ displate club membership.  😄 

Here is the log from the last startup with debug tag and changed env variables for extended logging:

[_firefly-firefly-app-1_logs.txt](https://github.com/firefly-iii/firefly-iii/files/12345277/_firefly-firefly-app-1_logs.txt)

**Comment by JC5** [comment]
Alright, the logs when you visit the home page, could you share a part of those? Feel free to email them instead. 

Ctrl-F "available()". That's the block of logs im curious about.

**Comment by JC5** [comment]
Thanks for sending the logs. I think I may have found the issue. Try running build nr. 591.`§

Since the issue is in the calculation, and the calculation is cached, try changing one budget amount (one euro up and down for example). I hope it will start showing correct amounts.

**Comment by Teonyr** [comment]
Unfortunately this fix didn't seem to work.
Logs send via email.

**Comment by JC5** [comment]
Thanks! One final idea I have is part of build 592. If you would be so kind. The `available()` logs have been slightly expanded as well. If it still doesn't work, I'd like to receive them.

Thanks for being part of my "live debug" session, I hope this is the final attempt!

**Comment by Teonyr** [comment]
Thx for the effort, but unfortunately I've send you an email.

**Comment by JC5** [comment]
Darn! Try 593!

**Comment by Teonyr** [comment]
It seems, my friend, this is the end of our conversation, since **you fixed it**. 😃 

Great work and thanks for helping me out. 👍🏻

**Comment by JC5** [comment]
Great! I'll mark it as fixed for 6.0.21 :+1:
